### PR TITLE
Fixed Bug with Training Set Split

### PR DESCRIPTION
**Pull request by hmathlee**
Fixed bug in code where external training set was being further divided into multiple folds instead of a separate train-val split. This caused complications with fine-tuning. Made clarifications on terminology between "trial" and "experiment" in comments and print lines.

### CODE REVIEW

File: src/external_validation.py [hunk 1]
```diff
@@ -379,124 +369,124 @@ def finetune_single_trial(self, trial_folder, hparams=None, lazy=True):
...
             # Create training and validation sets for fine-tuning.
-            k = self.k
-            val_num = len(train_df) // k
-            train_folds = []
-            i = 0
-            while i < len(train_df):
-                train_folds.append(train_df[i:i + min(len(train_df) - i, val_num)])
-                i += val_num
-            print("{} folds ready.".format(len(train_folds)))
+            sub_val_df = train_df.tail(int(len(train_df) * cfg['FINETUNE']['TRAIN_VAL']))
```
Comment (bvanberl) [inline src/external_validation.py:377]: Is looks like train_df isn't shuffled. Why not use train_test_split()?

Comment (hmathlee) [reply src/external_validation.py:377]: `train_df.sample(frac=1)` is a few lines above, which shuffles train_df.

File: src/external_validation.py [hunk 2]
```diff
@@ -321,16 +313,14 @@ def finetune_single_trial(self, trial_folder, hparams=None, lazy=True):
...
                   .format(str(num_folds_added + 1), len(fold_list)))
             # Evaluate the model on the external test set. Note that we load the original pre-finetuned model at each
             # iteration to minimize codependence between runs of a given trial.
             model = load_model(cfg_full['PREDICT']['MODEL'], compile=False)
             pred_labels, probs = predict_set(model, ext_df)
             # Can change these metrics as necessary
-            metrics = [Accuracy(name='accuracy'), AUC(), F1Score(num_classes=2, average='micro', threshold=0.5), Precision(name='precision'), Recall(name='recall'),
-                       TrueNegatives(), TruePositives(), FalseNegatives(), FalsePositives(), Specificity(name='specificity'),
-                       PhiCoefficient(), SensitivityAtSpecificity(0.95)]
+            metrics = [SensitivityAtSpecificity(0.95, name='sensitivity'), Specificity(name='specificity')]
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/external_validation.py:323]: Yes, good catch! Thanks

File: src/external_validation.py [hunk 3]
```diff
@@ -379,124 +369,124 @@ def finetune_single_trial(self, trial_folder, hparams=None, lazy=True):
...
+            if history is not None:
+                if len(history.epoch) < epochs:
+                    with writer2.as_default():
+                        tf.summary.text(name='Early Stopping', data=tf.convert_to_tensor('Training stopped early'),
+                                        step=0)
+
+            # for metric in history.history.keys():
+            #     if metric in metrics_df.columns:
+            #         metric_values = history.history[metric]
+            #         metrics_df.loc[metric] = sum(metric_values) / len(metric_values)
```
Comment (farazali7) [inline src/external_validation.py:469]: NP: remove if not needed

**Review by farazali7 — APPROVED** [review]
